kore/src/Kore/JsonRpc.hs

```diff
@@ -0,0 +1,270 @@
+import Control.Monad.Reader (ask, runReaderT)
+import Control.Monad.STM (atomically)
+import Data.Aeson.Types (FromJSON (..), ToJSON (..), Value (..), object, (.=))
+import Data.Conduit.Network (serverSettings)
+import Data.Text (Text)
+import Deriving.Aeson (
+    CamelToKebab,
+    CustomJSON (..),
+    FieldLabelModifier,
+    OmitNothingFields,
+    StripPrefix,
+ )
+import GHC.Generics (Generic)
+import Network.JSONRPC (
+    BatchRequest (BatchRequest, SingleRequest),
+    BatchResponse (BatchResponse, SingleResponse),
+    ErrorObj (..),
+    FromRequest (..),
+    JSONRPCT,
+    Request (..),
+    Respond,
+    Response (ResponseError),
+    Ver (V2),
+    buildResponse,
+    fromRequest,
+    jsonrpcTCPServer,
+    receiveBatchRequest,
+    sendBatchResponse,
+ )
+import Prelude.Kore
+
+data ExecuteRequest = ExecuteRequest
+    { state :: !Text
+    , maxDepth :: !(Maybe Int)
+    , haltPatterns :: ![Text]
+    }
+    deriving stock (Generic, Show, Eq)
+    deriving
+        (FromJSON)
+        via CustomJSON '[OmitNothingFields, FieldLabelModifier '[CamelToKebab]] ExecuteRequest
+
+newtype StepRequest = StepRequest
+    { state :: Text
+    }
+    deriving stock (Generic, Show, Eq)
+    deriving
+        (FromJSON)
+        via CustomJSON '[OmitNothingFields, FieldLabelModifier '[CamelToKebab]] StepRequest
+
+data ImpliesRequest = ImpliesRequest
+    { antecedent :: !Text
+    , consequent :: !Text
+    }
+    deriving stock (Generic, Show, Eq)
+    deriving
+        (FromJSON)
+        via CustomJSON '[OmitNothingFields, FieldLabelModifier '[CamelToKebab]] ImpliesRequest
+
+newtype SimplifyRequest = SimplifyRequest
+    { state :: Text
+    }
+    deriving stock (Generic, Show, Eq)
+    deriving
+        (FromJSON)
+        via CustomJSON '[OmitNothingFields, FieldLabelModifier '[CamelToKebab]] SimplifyRequest
+
+data ReqException = CancelRequest deriving stock (Show)
+
+instance Exception ReqException
+
+instance FromRequest (API 'Req) where
+    parseParams "execute" = Just $ fmap (Execute <$>) parseJSON
+    parseParams "step" = Just $ fmap (Step <$>) parseJSON
+    parseParams "implies" = Just $ fmap (Implies <$>) parseJSON
+    parseParams "simplify" = Just $ fmap (Simplify <$>) parseJSON
+    parseParams "cancel" = Just $ const $ return Cancel
+    parseParams _ = Nothing
+
+data PatternMatch = PatternMatch
+    { pmPattern :: !Int
+    , pmSubstitution :: !Text
+    }
+    deriving stock (Generic, Show, Eq)
+    deriving
+        (ToJSON)
+        via CustomJSON '[OmitNothingFields, FieldLabelModifier '[StripPrefix "pm", CamelToKebab]] PatternMatch
+
+{- HLINT ignore ReasonForHalting -}
+
+data ReasonForHalting
+    = HaltBranching
+        { depth :: !Int
+        }
+    | HaltStuck
+        { depth :: !Int
+        }
+    | HaltDepthBound
+    | HaltPatternMatch
+        { depth :: !Int
+        , matches :: ![PatternMatch]
+        }
+    deriving stock (Show, Eq)
```

Comment:
Won't this lead to partial record accessors? I'd avoid that if possible.

Reply:
Yes. What would you suggest instead? I wanted to keep close to the discussed spec, i.e.

```
{ "reason": "branching", "depth": DEPTH }
{ "reason": "stuck", "depth": DEPTH }
{ "reason": "depth-bound" }
{ "reason": "pattern-match", "depth": DEPTH, "matches": [ { "pattern": INT_INDEX_OF_PATTERN_THAT_MATCHED, "substitution": MATCHING_SUBSTITUTION } ] }
```